### ISSUE TITLE
rgb Image colors washed out 

### ISSUE THREAD

**Issue by tlambert03** [pygfx/pygfx#329]
Howdie,  and thanks for this very cool looking library... just getting my toes wet.

forgive the basic question, and apologies if this is discussed elsewhere, but I'm seeing washed out colors in the astronaut image when I follow the examples:
```python
# using either:..

image = gfx.Image(
    gfx.Geometry(grid=tex),
    gfx.ImageBasicMaterial(clim=(0, 255)),
)

image = gfx.Mesh(
    gfx.plane_geometry(512, 512),
    gfx.MeshBasicMaterial(map=tex.get_view(filter="linear")),
)
```

compared to vispy:

<img width="945" alt="Screen Shot 2022-09-20 at 3 04 45 PM" src="https://user-images.githubusercontent.com/1609449/191344865-498ec180-52f3-47fe-9398-4ec5d75e142d.png">

I'm on a mac (12.6), python 3.10, PyQt5 (though I also see it with jupyter-rfb in notebook).

curious if this would a lower OS-level driver-type issue? or something I can tackle within the pygfx code?

**Comment by Korijn** [comment]
Can you share more details about your environment? Versions of wgpu, pygfx and imageio would be helpful to see if we can reproduce this.

Thanks for reporting, we're still a young project so every bug report helps :)

**Comment by tlambert03** [comment]
thanks! i'm at:

pygfx              0.1.9
imageio            2.22.0
wgpu               0.8.1

<details>

```
Package            Version
------------------ -------
appnope            0.1.3
asttokens          2.0.8
backcall           0.2.0
cffi               1.15.1
decorator          5.1.1
executing          1.0.0
imageio            2.22.0
ipython            8.5.0
jedi               0.18.1
Jinja2             3.1.2
MarkupSafe         2.1.1
matplotlib-inline  0.1.6
numpy              1.23.3
parso              0.8.3
pexpect            4.8.0
pickleshare        0.7.5
Pillow             9.2.0
pip                22.2.2
prompt-toolkit     3.0.31
ptyprocess         0.7.0
pure-eval          0.2.2
pycparser          2.21
pygfx              0.1.9
Pygments           2.13.0
PySide6            6.3.2
PySide6-Addons     6.3.2
PySide6-Essentials 6.3.2
rubicon-objc       0.4.2
setuptools         65.3.0
shiboken6          6.3.2
six                1.16.0
stack-data         0.5.0
traitlets          5.4.0
wcwidth            0.2.5
wgpu               0.8.1
wheel              0.37.1
```

</details>

to reproduce

```sh
conda create -n test python=3.10
conda activate test
pip install pygfx pyside6 imageio
```

```python
import imageio
from wgpu.gui.auto import WgpuCanvas, run
import pygfx as gfx

canvas = WgpuCanvas()
renderer = gfx.renderers.WgpuRenderer(canvas)
scene = gfx.Scene()

im = imageio.imread("imageio:astronaut.png")
image = gfx.Image(
    gfx.Geometry(grid=gfx.Texture(im, dim=2)),
    gfx.ImageBasicMaterial(clim=(0, 255)),
)
scene.add(image)

camera = gfx.OrthographicCamera(512, 512)
camera.position.set(256, 256, 0)
camera.scale.y = -1

if __name__ == "__main__":
    canvas.request_draw(lambda: renderer.render(scene, camera))
    run()
```


if it's not a "known" issue with an obvious solution (i.e., something I'm just doing wrong), I'm happy to dig a little deeper

**Comment by tlambert03** [comment]
additional data point: I booted windows on the same hardware (dual boot on a different drive) and got the same result.  so it doesn't appear to be mac-specific

**Comment by almarklein** [comment]
Could be related to the [srgb colorspace](https://en.wikipedia.org/wiki/SRGB).

**Comment by Korijn** [comment]
Silly question: are the image files the same?

**Comment by tlambert03** [comment]
> Could be related to the [srgb colorspace](https://en.wikipedia.org/wiki/SRGB). *edit:* nope, that does not seem to fully explain it.

It does _feel_ like a colorspace issue (even if you've already concluded it's not that 😊)

**Comment by almarklein** [comment]
@Korijn yeah, I checked against mpl (using the same image) and I get the same result.

Indeed I cannot imagine what else it could be. Double-checked some shader code, but see nothing suspicious. If I turn off the srgb format, the images look more similar, but still not the same. Maybe gamma correction?

I think I'll try showing a few squares in pure black, white, red, green and blue, and see if these colors doe match. Then we know whether the problem is in the image shader or in the final composition step or something.

**Comment by almarklein** [comment]
Drawing the color #333333, a.k.a. `(0.2, 0.2, 0.2, 1.0)` a.k.a. `(51, 51, 51, 255)`:

* With the wgpy triangle example (as bare bones as it gets), with srgb target texture: 124
* Dito, but with `bgra8unorm`: 50
* That color in pygfx: same results.
* In MPL: 66

So looks like MPL does *something* but it's not quite srgb?

**Comment by tlambert03** [comment]
fascinating

**Comment by Korijn** [comment]
So _none_ of them result in 51?

**Comment by almarklein** [comment]
I can get it very close to MPL's result by not using the srgb target, but applying a gamma correction of about 0.82. It's within 2 or 3 values for my test image (have not yet computationally compared the astronaut image).

<img width="521" alt="image" src="https://user-images.githubusercontent.com/3015475/191790674-fb9b699b-5c4e-4403-842d-243936004a65.png">

**Comment by almarklein** [comment]
Some more observations:

### Color value rounding

Running from `0.0`  to `1.0`  we get something like this:
![image](https://user-images.githubusercontent.com/3015475/191946243-aa3f9c28-2b12-419d-bb3a-73f4b394e8dd.png)

![image](https://user-images.githubusercontent.com/3015475/191946152-d823b1a0-f6d8-420d-9eb8-8dafb65c5a97.png)

You can see how the regions that gets 0 and 255 are smaller. This implies that conversion from float `[0..1]` is done with `round(val*255)`. This is to be expected, and what we use in th pygfx Color object. Nothing odd here, but good to have this verified.

### 51 does not exist?

Because `0.2 * 255 = 51.0`, we'd expect `#333` to result in 51. But as we saw above it does not. So I modified the wgpu-py triangle example to interpolate between a bit less and a bit more than 0.2. And then ... the value skips from 50 to 52. It seems impossible to get the value 51 on screen! 🤣  🤷 

### The other values are ok

I also checked `#666` , `#999` and `#ccc`, and things are fine for these:
![image](https://user-images.githubusercontent.com/3015475/191949047-13655c7c-4a0f-49e8-ace0-523b4392bda5.png)

This means that with wgpu we can display colors exactly as they are (except for that magical #333).

**Comment by Korijn** [comment]
I think this slide is very reminiscent of the situation we are looking at here:

![image](https://user-images.githubusercontent.com/1882046/191952730-83e91c8d-619f-4ea2-aa3e-05ca5e766c87.png)

Except we appear to not be applying the gamma correction in our code.

(Source: http://filmicworlds.com/blog/linear-space-lighting-i-e-gamma/)

**Comment by almarklein** [comment]
That post has made things more clear for me. Thanks!

So what's been happening: 
```
image  on hd                                                  using srgb final target texture for good 
in gamma space   ->   we do stuff       ->      appearance
(brighter)                     maybe lighting
```

The image as stored on the HD is prepared to be send to the HD directly. But we do as second gamma correction using the srgb target, so it appears too bright.

I also learned that srgb is very similar to gamma 1/2.2, with some small tweaks related to how the human eye observes colors.

The solution should be that we DO use the srgb target texture, but also turn images/textures back to linear space using `pow(val, 2.2)`.

**Comment by almarklein** [comment]
Important question: how do we know whether an image is in gamma space? 🤔

**Comment by Korijn** [comment]
> Important question: how do we know whether an image is in gamma space? 🤔
> 
> I guess we should assume that it is, and provide an option to turn it off ...

I think you just assume the "industry standard" value unless an image provides gamma metadata: https://en.wikipedia.org/wiki/Gamma_correction#Gamma_meta_information

**Comment by almarklein** [comment]
Yeah, but by the time the image reached pygfx the image is just a numpy array, and we don't know whether this was a png, jpeg, raw,  generated data, or whatever.

So either
* We assume that it was an image that is in gamma space, so people would have to set a property/argument if their data is not.
* We assume that all data is in linear space, and provide a function `ungamma()` that must be used for every image read with imageio. We'd have to educate our users on using it, but if we have it in all examples, maybe that works?

**Comment by Korijn** [comment]
I think the first proposal makes the most common scenario easiest?

**Comment by almarklein** [comment]
mmm ...

I started with specifying gamma correction (to bring to linear space) in the texture. But texture data can be anything, 2D images is one thing, but there are also 1D colormaps, 3D translation data, whatever. Not a good place to apply a default power transform.

Maybe in ImageVisual. But there are more places where image data can enter into pygfx. E.g. as a 2D colormap on a mesh.

Since image data can be used in pygfx in different places, and it would be confusing if some places would auto-convert while others  would not, I think the transform must be made when the data is read.

My proposal would be to:
* Introduce `pygfx.gamma_transform()`.
* Maybe introduce `pygfx.imread(uri, gamma_correction=2.2)`, that uses `imageio.imread()` and `gamma_transform()`.
* Or ... contribute to imageio to make the `ignoregamma` that exists for png, available for all formats.

What a mess, this gamma stuff! I just realized that if you read a png and process it with e.g. skimage, you're probably processing the wrong data. It does not appear to me that this is widely known ...

**Comment by almarklein** [comment]
The option to make imageio more suited for reading the "original" values seems to be the most reliable, since there the meta data is available.

**Comment by almarklein** [comment]
I'm confused 🤔 

I replicated the test visualization for MPL using an image instead of `Rectangle` objects. It yields the exact same results. This means that MPL does not threat the image different from other visuals. In other words, it does not "undo the builtin gamma-transform in the image".

Another thing. Given that the below squares have values at a fixed interval, and the gradient is linear (in terms of values) which of the two images *looks* more linear?
<img width="636" alt="image" src="https://user-images.githubusercontent.com/3015475/192065534-c457bcb2-d838-46c4-a963-b58a6e3984aa.png">

<img width="633" alt="image" src="https://user-images.githubusercontent.com/3015475/192065505-32d15135-1b4d-437f-990a-34b77c2e242f.png">


On the three screens that I tried, its definitely the bottom one - which is the one without the final srgb conversion. And you can pick the values and see that they're *still* linear. WAT?

**Comment by Korijn** [comment]
Does MPL perhaps do some tricks in imread?

**Comment by almarklein** [comment]
> Does MPL perhaps do some tricks in imread?

I was using imagio in both cases.

> Also... Does picking yield colors before or after screen manipulation?

I take a screenshot and pick in an image editor. The screen's framebuffer is grabbed - what's fed to the monitor.

After a night's sleep, I understand it a bit better, I think. The human eye's perception of light is not linear. So linear pixel values should *not* result in a linear appearing gradient. That is, not if you want to render realistic looking stuff. If you render lines and stuff, all this matters much less.

**Comment by panxinmiao** [comment]
Sorry, I may have missed this.

I'm sure it's a matter of "color space". "srgb space" is used when rendering to the screen, while the original image works in "linear space". I have encountered this problem in this PR: #324 , and this PR has solved this problem to some extent,  The `gfx. Texture`  now has an `encoding` attribute, which needs to be set to 'srgb' to make it work in the "srgb space".

I did a test (based on #324)

```
import imageio
from wgpu.gui.auto import WgpuCanvas, run
import pygfx as gfx

canvas = WgpuCanvas()
renderer = gfx.renderers.WgpuRenderer(canvas)
scene = gfx.Scene()

im = imageio.imread("imageio:astronaut.png")
image = gfx.Image(
    gfx.Geometry(grid=gfx.Texture(im, dim=2, encoding='srgb')),
    gfx.ImageBasicMaterial(clim=(0, 255)),
)
scene.add(image)

camera = gfx.OrthographicCamera(512, 512)
camera.position.set(256, 256, 0)
camera.scale.y = -1

if __name__ == "__main__":
    canvas.request_draw(lambda: renderer.render(scene, camera))
    run()
```

![image](https://user-images.githubusercontent.com/8044566/192205260-53404125-ce37-496c-9b15-b941ce3398ed.png)

**Comment by Korijn** [comment]
I looked through your changes and noticed you are configuring the [texture format](https://webgpu.rocks/reference/enum/gputextureformat/#idl-gputextureformat) to include a `-srgb` suffix.

I guess this triggers a different texture sampling mode?

**Comment by panxinmiao** [comment]
> I looked through your changes and noticed you are configuring the [texture format](https://webgpu.rocks/reference/enum/gputextureformat/#idl-gputextureformat) to include a `-srgb` suffix.
> 
> I guess this triggers a different texture sampling mode. wgpu docs on the subject here: https://github.com/gfx-rs/wgpu/wiki/Texture-Color-Formats-and-Srgb-conversions
> 
 Yes, I think the wgpu made internal color space conversion.

I also encountered a similar problem when implementing PBR rendering. The color of the scene is too bright and distorted. I learned the reason behind it from here: https://learnopengl.com/Advanced-Lighting/Gamma-Correction. (It's really a great tutorial!)

**Comment by almarklein** [comment]
After reading [what-every-coder-should-know-about-gamma](https://blog.johnnovak.net/2016/09/21/what-every-coder-should-know-about-gamma/) My current (ongoing) understanding is that about every color that goes into pygfx can be expected to be in srgb space. Designers define/draw colors in srgb. Web colors are srgb, and browsers basically assume all images to be in srgb. This also makes sense, because to store a color in 8 bits per channel, srgb is much more effective because the human eye is nonlinear; srgb colors are linear from the point of view of human perception. 

You can blissfully ignore all this, and send the (srgb) colors to the final texture, and all is well, because the final texture (i.e. the monitor) needs to get srgb colors. Except if you do calculations with these colors, which are not linear in the physical sense (amount of photons), so you'd get odd results.

So what we need to do (what I currently think), is in the shader tranform each color to physical space, do all our calculations, and *eventually* transform the color back to srgb. These transformations can be done using a simple power function, or using srgb textures.

I have a few small open questions. I'll dig a bit more and will try to write things up as I make updates to pygfx (might become a blog post).

**Comment by panxinmiao** [comment]
> Well, that fixed it for that example. I I think I want to propose to convert each color value in the shader. That way the user won't have to worry about it in general. Plus then it also works for basic materials, lines etc. I think we'd only need the srgb textures for render targets.
> 
In fact, when using a texture in the srgb format (that is, the texture is interpreted as working in the srgb space), the shader program will automatically perform the conversion from the srgb space to the linear space when sampling from the texture. This part is hardware accelerated and does not need to be processed in the shader.

Similarly, when outputting to the screen, that is, when the shader output is written to the srgb format texture, the conversion from linear space to srgb space will also be performed automatically.

**Comment by almarklein** [comment]
That's correct, but I think for us the use for sampling from an srgb texture is limited. In the mesh shader the color can come from several places. E.g. a uniform color, per-vertex colors, or via a colormap. It's much easier to apply the gamma conversion in the shader once we have the color. Implemented this now for all our shaders in #330.

**Comment by tlambert03** [comment]
It appears so! 😂

I've learned a lot as well by reading along, so I'm glad I posted it :)
